From decacd193e4bfab4cd5b573200836d6e68c10083 Mon Sep 17 00:00:00 2001
From: Paul Keith <javelinanddart@gmail.com>
Date: Sat, 10 Mar 2018 05:16:20 +0100
Subject: [PATCH 1/3] Telephony: Fallback gracefully for emergency calls if
 suitable app isn't found

* We can safely fall back to Google's if the user-chosen app doesn't
  support emergency calls, and if Google's is present on the device
* Show an alert dialog to the user if no suitable app is found
  for handling emergency calls instead of bringing down the system

Change-Id: Ieda12656df7ddc0a9f7d4630f5c99a8596488908
---
 res/values/cm_arrays.xml                      | 12 +++++++
 res/values/cm_strings.xml                     |  3 ++
 .../phone/OutgoingCallBroadcaster.java        | 31 +++++++++++++------
 3 files changed, 37 insertions(+), 9 deletions(-)

diff --git a/res/values/cm_arrays.xml b/res/values/cm_arrays.xml
index 437f7ef2e..75cf2bef5 100644
--- a/res/values/cm_arrays.xml
+++ b/res/values/cm_arrays.xml
@@ -40,4 +40,16 @@
         <item>@string/preferred_network_mode_gsm_only_choice</item>
         <item>@string/preferred_network_mode_gsm_wcdma_preferred_choice</item>
     </string-array>
+
+    <!-- Package names for dialer apps that can handle emergency calls -->
+    <string-array name="trusted_dialer_packages" translatable="false">
+        <item>@string/ui_default_package</item>
+        <item>com.android.google.dialer</item>
+    </string-array>
+
+    <!-- Class names of main activities for dialer apps that can handle emergency calls -->
+    <string-array name="trusted_dialer_classes" translatable="false">
+        <item>@string/dialer_default_class</item>
+        <item>com.android.google.dialer.extensions.GoogleDialtactsActivity</item>
+    </string-array>
 </resources>
diff --git a/res/values/cm_strings.xml b/res/values/cm_strings.xml
index 23d87e9f7..f75c966a9 100644
--- a/res/values/cm_strings.xml
+++ b/res/values/cm_strings.xml
@@ -36,4 +36,7 @@
     <string name="preferred_network_mode_wcdma_only_choice">WCDMA only</string>
     <string name="preferred_network_mode_gsm_only_choice">GSM only</string>
     <string name="preferred_network_mode_gsm_wcdma_preferred_choice">GSM/WCDMA preferred</string>
+
+    <!-- Emergency Calling -->
+    <string name="emergency_unsupported_dialer_message">Unable to complete emergency call, no suitable app found</string>
 </resources>
diff --git a/src/com/android/phone/OutgoingCallBroadcaster.java b/src/com/android/phone/OutgoingCallBroadcaster.java
index 6bb138897..33e683247 100644
--- a/src/com/android/phone/OutgoingCallBroadcaster.java
+++ b/src/com/android/phone/OutgoingCallBroadcaster.java
@@ -474,15 +474,28 @@ public class OutgoingCallBroadcaster extends Activity
                 // opposed to a Context which look up known android
                 // packages only)
                 final Resources resources = getResources();
-                invokeFrameworkDialer.setClassName(
-                        resources.getString(R.string.ui_default_package),
-                        resources.getString(R.string.dialer_default_class));
-                invokeFrameworkDialer.setAction(Intent.ACTION_DIAL);
-                invokeFrameworkDialer.setData(intent.getData());
-                if (DBG) Log.v(TAG, "onCreate(): calling startActivity for Dialer: "
-                               + invokeFrameworkDialer);
-                startActivity(invokeFrameworkDialer);
-                finish();
+                String[] packages = resources.getStringArray(R.array.trusted_dialer_packages);
+                String[] classes = resources.getStringArray(R.array.trusted_dialer_classes);
+                for (int i = 0; i < packages.length; i++) {
+                    try {
+                        invokeFrameworkDialer.setClassName(packages[i], classes[i]);
+                        invokeFrameworkDialer.setAction(Intent.ACTION_DIAL);
+                        invokeFrameworkDialer.setData(intent.getData());
+                        if (DBG) Log.v(TAG, "onCreate(): calling startActivity for Dialer: "
+                                       + invokeFrameworkDialer);
+                        startActivity(invokeFrameworkDialer);
+                        finish();
+                        return;
+                    } catch (IllegalArgumentException e) {
+                        continue;
+                    }
+                }
+
+                // No dialer was able to handle the emergency call
+                new AlertDialog.Builder(this)
+                        .setMessage(R.string.emergency_unsupported_dialer_message)
+                        .setPositiveButton(android.R.string.ok, null)
+                        .show();
                 return;
             }
             callNow = false;
-- 
2.17.1

